### DISCUSSION TITLE
Maximal weight matching

### DISCUSSION THREAD

**Discussion by VukMNE** [networkx/networkx#5581] [Q&A] [answered]
Hello everyone,

I am not sure if I discovered a potential bug in finding the maximal weight matching, or if I am not seeing the bug in my own code. Either way I will start a discussion here, and see how it goes.

 As a project for some Msc Data Science programme, I worked on searching for a maximal weight matching on 20x20 graph grid. I used [networkx.max_weight_matching](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.matching.max_weight_matching.html) method to find it. 

Today, I wrote my custom method, which only adds eligible edge with the highest weight to the matching. Surprisingly, that method returns a matching that has **higher sum of weights than networkx.max_weight_matching**, and it had more edges within the matching also. 

```
def maximal_weight_matching_local_search(G, w):
    # First let's assign weights to the graph
    g = G.copy()
    for i, e in enumerate(g.edges):
        g[e[0]][e[1]]['weight'] = w[i]

    matching = set()
    remaining_edges = list(g.edges)
    added_nodes = set()
    while len(remaining_edges) > 0:
        # find the edge with maximum weight
        rem_weights = [g[e[0]][e[1]]['weight'] for e in remaining_edges]
        max_weight_pos = np.argmax(rem_weights)
        max_edge = remaining_edges[max_weight_pos]
        matching.add(max_edge)
        added_nodes.add(max_edge[0])
        added_nodes.add(max_edge[1])
        g.remove_edge(max_edge[0], max_edge[1])
        remaining_edges = list(g.edges)
        remaining_edges = list(filter(lambda edge: edge[0] not in matching and edge[1] not in added_nodes, remaining_edges))

    return matching
```

I understand that my code is silly, and that it would fail for a simple graph with weights like this one:
```
+--5--+
|     |
4     4
|     |
+--1--+
```

But how come that I was able to get a better result than with the networkx method?

Below, I provide the code with which you can try to verify if everything is correct.

```
G = nx.grid_graph([20, 20])
weights = np.random.uniform(1, 2, len(G.edges))

# let's assign weights to the graph
for i,e in enumerate(G.edges):
    G[e[0]][e[1]]['weight'] = w[i]

nx_max = nx.max_weight_matching(G, maxcardinality=False, weight='weight') 
my_max = maximal_weight_matching_local_search(G, w)

nx_sum= 0
for e in nx_max:
    nx_sum+= G[e[0]][e[1]]['weight']

my_sum = 0
for e in my_max :
    my_sum += G[e[0]][e[1]]['weight']

print('NX sum:' + str(nx_sum))
print('my sum:' + str(my_sum))
```

**Comment by dschult** [comment, marked as answer]
It looks like your method doesn't return a matching:
```
>>> nx.is_matching(my_max)
False
>>> nx.is_matching(nx_max)
True
```
I think that might be because the `lambda` function you define in your filter is checking if `edge[0] not in matching` instead of `edges[0] not in added_nodes`.

**Reply by VukMNE** [reply]
Yes, you are correct. I am sorry for the confusion
